### PR TITLE
refactor(seismic): migrate e2e integration tests to in-process node

### LINKED ISSUE #338: Migrate seismic e2e tests from subprocess to in-process node harness
## Context

PR #336 fixed immediate lifecycle issues (orphaned processes, port conflicts, retry storms) in the seismic integration test. However, the underlying test architecture still shells out to `cargo run --bin seismic-reth`, which is fundamentally brittle and slow.

## Current problems

1. **Recompilation at test time** — `cargo run --bin seismic-reth` reinvokes the compiler inside the test (`utils.rs:47`). Even though the test binary itself was just compiled, the `cargo run` triggers a separate compilation of the `seismic-reth` binary, adding significant time especially in CI.
2. **Hardcoded HTTP RPC port** — `url()` returns `127.0.0.1:8545` (`utils.rs:132`), a latent source of conflicts if anything else binds that port.
3. **Stdout parsing for readiness** — the test waits for `"Starting consensus engine"` in stdout to know the node is up; any log format change silently breaks this.
4. **Silent hangs on read errors** — `read_line` errors in the spawned stdout/stderr tasks panic silently (the `JoinHandle` is never awaited), causing the test to hang until the nextest timeout.
5. **No isolation** — all tests share the same filesystem paths and ports; parallel execution is impossible.

## Proposed improvement

Upstream reth provides `reth_e2e_test_utils` which starts a full node **in-process** via `NodeBuilder::new(config).testing_node(executor)`. Migrating to this pattern would:

- **Eliminate recompilation** — the node is compiled once as part of the test binary, not again at runtime via `cargo run`.
- **Remove port conflicts** — RPC is in-process, no TCP ports needed.
- **Improve isolation** — each test gets its own `TempDatabase`.
- **Give direct access** — call into node internals (pool, provider, payload builder) instead of going through HTTP RPC.
- **Simplify lifecycle** — no process management, stdout parsing, or `kill_on_drop`.

The target API would look like:

```rust
let (mut nodes, _tasks, wallet) = setup::<SeismicNode>(
    1, Arc::new(chain_spec), false, eth_payload_attributes,
).await?;
let mut node = nodes.pop().unwrap();
let raw_tx = TransactionTestContext::transfer_tx_bytes(1, wallet.inner).await;
let tx_hash = node.rpc.inject_tx(raw_tx).await?;
let payload = node.advance_block().await?;
```

This may require implementing additional traits on `SeismicNode` to satisfy the `NodeBuilder` test infrastructure.

## Intermediate step

Before the full migration, the hardcoded port 8545 can be addressed by parsing the bound port from reth's startup logs and passing it to `url()` dynamically.

## References

- PR #336 (lifecycle fixes): https://github.com/SeismicSystems/seismic-reth/pull/336
- Upstream e2e pattern: `crates/ethereum/node/tests/e2e/eth.rs`
- Current test harness: `crates/seismic/node/src/utils.rs`

### PR DESCRIPTION
**Pull request by ameya-deshmukh**
## Summary

Replaces the subprocess-based integration test architecture (`cargo run --bin seismic-reth`) with in-process node testing via `setup(1)` + `advance_block()`, matching the pattern used by upstream reth and already adopted by our `p2p.rs`, `testsuite.rs`, and `hardfork_config.rs` tests.

- **Split monolithic `integration_test` into 9 independent `#[tokio::test]` functions** — each spawns its own in-process node on random ports, enabling parallel execution and individual failure reporting
- **Move Seismic RPC registration into `SeismicAddOns::launch_add_ons`** — `EthApiExt` (eth_ overrides) and `SeismicApi` (seismic_ namespace) are now registered as part of the node's add-ons lifecycle instead of via `extend_rpc_modules` in `main.rs`. This ensures both the binary and in-process test nodes get the full Seismic RPC surface.
- **Remove `SeismicRethTestCommand`** — the subprocess launcher (`cargo run --bin seismic-reth`) and its stdout-parsing readiness detection are no longer needed
- **Extract shared async helpers** to keep async state machines small (`rpc_test_deploy_contract`, `flagged_storage_deploy_and_write`, etc.)
- **Heap-allocate node setup via `tokio::spawn`** to avoid stack overflow on x86_64 CI (see below)
- **Re-export `FullSeismicApi`** from `reth_seismic_rpc::eth` module so `pub use eth::*` pulls it in cleanly

## Stack overflow on x86_64 CI

The in-process tests overflow the default 2MB thread stack on x86_64 Linux CI, but not on ARM64 macOS locally.

**Root cause**: Rust compiles each `async fn` into a state machine enum where every `.await` point becomes a variant holding all locals live across that point. The test futures are large because they hold `SeismicTestNode` (contains the entire running node's state — provider, pool, network, RPC registry) plus sub-futures for RPC calls, all composed into a single enum. The total future size + the call chain depth (tokio runtime → test state machine → jsonrpsee client → jsonrpsee server → Seismic RPC handler → EVM execution) exceeds 2MB on x86_64.

**Why x86_64 but not ARM64**: x86_64 has 16 general-purpose registers vs ARM64's 31. Fewer registers means more local variables spill to the stack, making each function's stack frame ~30-50% larger. The same Rust code that fits in 2MB on ARM64 overflows on x86_64 in debug builds.

**Fix**: `tokio::spawn(setup(1))` in `setup_test_node()` moves the large `setup` future from the caller's stack onto the heap. `tokio::spawn` boxes the future internally (`Box::pin`) and runs it on a worker thread as a standalone task — the caller only holds a lightweight `JoinHandle` on their stack. This eliminates the need for `RUST_MIN_STACK` entirely and is robust against future growth in test logic.

## What changed

| File | Change |
|------|--------|
| `crates/seismic/node/tests/e2e/integration.rs` | 9 independent in-process tests with extracted helpers; `tokio::spawn` for node setup |
| `crates/seismic/node/src/node.rs` | Register `EthApiExt` + `SeismicApi` in `launch_add_ons` |
| `crates/seismic/node/src/utils.rs` | Remove `SeismicRethTestCommand` |
| `crates/seismic/rpc/src/lib.rs` | Simplified to `pub use eth::*` (FullSeismicApi re-exported from eth/mod.rs) |
| `crates/seismic/rpc/src/eth/mod.rs` | Added `pub use api::FullSeismicApi` |
| `bin/seismic-reth/src/main.rs` | Remove redundant `extend_rpc_modules` (now in `launch_add_ons`) |
| `crates/seismic/node/Cargo.toml` | Remove unused deps (`once_cell`, `tempfile`), make `serde_json` dev-only |
| `.github/workflows/seismic.yml` | Removed `RUST_MIN_STACK` (no longer needed) |
| `.config/nextest.toml` | Remove stale port-conflict override |

## Why

Issue #338 identified that the subprocess test architecture was:
1. **Slow** — `cargo run` reinvokes the compiler at test time
2. **Brittle** — hardcoded port 8545, stdout parsing for readiness
3. **Non-isolated** — all tests shared one node, no parallel execution

The in-process approach eliminates all three: the node compiles once as part of the test binary, each test gets its own ephemeral database and random ports, and blocks are produced deterministically via `advance_block()` instead of `thread::sleep`.

## Test plan

- [x] `cargo nextest run -p reth-seismic-node -E 'kind(test)' --no-fail-fast` — all 15 tests pass locally (no `RUST_MIN_STACK` needed)
- [x] `cargo check -p reth-seismic-rpc -p reth-seismic-node` — compiles clean
- [x] `cargo clippy` with CI flags — no errors
- [x] `cargo check -p seismic-reth` — binary still compiles
- [x] `cargo +nightly fmt --all --check` — formatted

Closes #338

### CODE REVIEW

File: crates/seismic/rpc/src/lib.rs [hunk 1]
```diff
@@ -9,7 +9,7 @@
 #![cfg_attr(docsrs, feature(doc_cfg, doc_auto_cfg))]
 
 mod eth;
-pub use eth::*;
+pub use eth::{api::FullSeismicApi, *};
```
Comment (samlaf) [inline crates/seismic/rpc/src/lib.rs]: do we want to `pub use api::FullSeismicApi;` in eth module so `*` also pulls that in? Or is there a reason why you didn't do that?

File: bin/seismic-reth/src/main.rs [hunk 2]
```diff
@@ -18,27 +16,12 @@ fn main() {
         // Boot enclave and fetch purpose keys BEFORE building node components
         let purpose_keys = boot_enclave_and_fetch_keys(&encl).await;
 
-        // Store purpose keys in global static storage before building the node
-        reth_seismic_node::purpose_keys::init_purpose_keys(purpose_keys.clone());
+        // Store purpose keys in global static storage before building the node.
+        // Seismic RPC modules (seismic_ namespace + eth_ overrides) are registered
+        // in SeismicAddOns::launch_add_ons, which reads keys via get_purpose_keys().
+        reth_seismic_node::purpose_keys::init_purpose_keys(purpose_keys);
```
Comment (samlaf) [inline bin/seismic-reth/src/main.rs:22]: completely orthogonal to this PR, but I feel like there has to be a better way to manage these keys, and they should prob be injected instead of being read via a global side channel like this. Claude seems to agree:

```
What it does: On boot, the node contacts the TEE enclave (or a mock) to fetch cryptographic "purpose keys" (used for ECDH decryption of TxSeismic calldata, etc.). These keys are   
stashed in a static OnceLock and later read by node.rs when registering the Seismic RPC endpoints.
                                                                                                                                                                                    
Is there a better way? The OnceLock global is a pragmatic escape hatch, but it's not great because:                                                                                 
- It couples initialization order (enclave boot → init_purpose_keys → node launch) through implicit global state
- It panics at runtime if the order is wrong, rather than enforcing it at compile time                                                                                              
- It makes testing harder (the ensure_mock_purpose_keys in test setup is a symptom)   
                                                                                                                                                                                    
A cleaner approach would be to inject the keys structurally — e.g., store them in SeismicNode itself:                                                                               
                                                                                                                                                                                    
// Instead of SeismicNode::default()                                                                                                                                                
let node = SeismicNode::new(purpose_keys);                                                                                                                                          
// Then SeismicAddOns can access them via self, no global needed
                                                                                                                                                                                    
This way the compiler enforces that keys exist before launch, and tests can construct nodes with mock keys directly without touching global state. It would require threading the   
keys through SeismicAddOns (which already has access to self in launch_add_ons_with), but that's the idiomatic reth pattern — data flows through the node/addons types, not globals.
```

File: crates/seismic/node/src/node.rs [hunk 3]
```diff
@@ -306,6 +309,16 @@ where
...
+    EthB::EthApi: reth_seismic_rpc::FullSeismicApi + Send + Sync + 'static,
+    <EthB::EthApi as reth_rpc_eth_api::EthApiTypes>::Error: Send + Sync + 'static,
+    jsonrpsee::types::ErrorObject<'static>:
+        From<<EthB::EthApi as reth_rpc_eth_api::EthApiTypes>::Error>,
+    <<EthB::EthApi as reth_rpc_eth_api::EthApiTypes>::NetworkTypes as reth_rpc_eth_api::RpcTypes>::TransactionRequest:
+        From<alloy_rpc_types::TransactionRequest>
+            + AsRef<alloy_rpc_types::TransactionRequest>
+            + Send
+            + Sync
+            + 'static,
```
Comment (samlaf) [inline crates/seismic/node/src/node.rs:321]: also orthogonal to this PR but this is super verbose and makes it hard to reason about what we're actually doing. Here's claude's take:
```
 Are they all needed? Yes, mechanically — the compiler requires them because EthApiExt::new(eth_api).into_rpc() needs to convert the eth API into a jsonrpsee RPC module, which
demands all those trait impls.

Could it be cleaner? Yes. The verbose associated-type-of-associated-type bounds (especially that TransactionRequest line) are a symptom of leaking internal implementation details
into the NodeAddOns impl. Two approaches to simplify:

1. A supertrait alias — define a single trait in reth_seismic_rpc that bundles all the RPC-related bounds:
pub trait SeismicRpcCompatible: FullSeismicApi + Send + Sync + 'static
where
    Self::Error: Send + Sync + 'static,
    ErrorObject<'static>: From<Self::Error>,
    // ...
{}
1. Then the impl block just says EthB::EthApi: SeismicRpcCompatible.
2. Concrete types instead of generics — upstream stays generic over EthB for extensibility, but Seismic always uses SeismicEthApiBuilder. If you don't need that flexibility,
collapsing the generics removes most bounds entirely.

The current version is correct but noisy. A trait alias would be the least invasive cleanup.
```

File: .github/workflows/seismic.yml [hunk 4]
```diff
@@ -103,6 +103,9 @@ jobs:
     timeout-minutes: 30
     env:
       SEISMIC_CI: 1
+      # In-process e2e tests have large async futures (node + engine + RPC + EVM).
+      # x86_64 stack frames are larger than ARM64; default 2MB overflows on CI.
+      RUST_MIN_STACK: 8388608
```
Comment (samlaf) [inline .github/workflows/seismic.yml]: nit: prob can be "solved" properly by boxing or something:

```
1. Box::pin the heavy futures — wrapping the largest async calls in Box::pin(async { ... }) moves their state from the caller's stack frame to the heap. This is the most targeted
fix:
let result = Box::pin(setup(1)).await?;
2. #[async_recursion] or manual boxing in setup() — if the deep nesting is inside the node launch internals, boxing there once fixes all callers.
3. Reduce the node surface — the in-process test node launches the full stack (engine, all RPC namespaces, networking, etc.). A lighter test-only node configuration that only
starts what tests actually need (HTTP RPC + EVM + block production) would have smaller futures. Upstream's reth_e2e_test_utils is designed for full nodes though, so this would be a
 bigger lift.
4. Use tokio::spawn to isolate setup — spawning the node setup on a separate task moves its entire future to the heap:
let (node, client, ...) = tokio::spawn(setup_test_node()).await??;

Option 1 or 4 would let you drop the RUST_MIN_STACK env var entirely with minimal code changes. The current approach (8MB stack + splitting into helpers) works but is fragile —
adding more logic to a test function can silently re-trigger the overflow.
```

**Review by samlaf — APPROVED** [review]
LGTM, very nice cleanup! Left a bunch of nits, most of which are orthogonal to this PR and should be fixed independently in future PRs imo.

File: crates/seismic/node/tests/e2e/integration.rs [hunk 5]
```diff
@@ -387,67 +324,43 @@ async fn test_seismic_reth_rpc() {
...
 }
 
-async fn test_seismic_reth_rpc_with_typed_data() {
-    let reth_rpc_url = SeismicRethTestCommand::url();
-    let chain_id = SeismicRethTestCommand::chain_id();
-    let client = jsonrpsee::http_client::HttpClientBuilder::default().build(reth_rpc_url).unwrap();
-    let wallet = Wallet::default().with_chain_id(chain_id);
+#[tokio::test(flavor = "multi_thread")]
+async fn test_seismic_reth_rpc() -> eyre::Result<()> {
+    let (mut node, client, chain_id, wallet, _tasks) = setup_test_node().await?;
```
Comment (cdrappi) [inline crates/seismic/node/tests/e2e/integration.rs:337]: wow this is so much better

**Review by cdrappi — APPROVED** [review]

File: crates/seismic/rpc/src/lib.rs (hunk 1 above)
Comment (ameya-deshmukh) [reply crates/seismic/rpc/src/lib.rs]: Done — added `pub use api::FullSeismicApi;` to `eth/mod.rs` and simplified `lib.rs` to just `pub use eth::*;`. dc6e1c9f2

File: .github/workflows/seismic.yml (hunk 4 above)
Comment (ameya-deshmukh) [reply .github/workflows/seismic.yml]: Resolved — wrapped `setup(1)` in `tokio::spawn` to move the large async state machine onto the heap, and removed the `RUST_MIN_STACK` env var from CI. dc6e1c9f2

File: bin/seismic-reth/src/main.rs (hunk 2 above)
Comment (ameya-deshmukh) [reply bin/seismic-reth/src/main.rs:22]: Opened #360 to address this.

File: crates/seismic/node/src/node.rs (hunk 3 above)
Comment (ameya-deshmukh) [reply crates/seismic/node/src/node.rs:321]: Opened #361 to address this.
